src/main/java/com/iota/iri/service/garbagecollector/GarbageCollector.java

```diff
@@ -0,0 +1,323 @@
+     * back into the corresponding job.
+     *
+     * @param jobClass class of the job that the GarbageCollector shall be able to handle
+     * @param jobParser parser function for the serialized version of jobs of the given type
+     */
+    void registerParser(Class<?> jobClass, JobParser jobParser) {
+        this.jobParsers.put(jobClass.getCanonicalName(), jobParser);
+    }
+
+    /**
+     * This method allows us to register a {@link QueueProcessor} for the given job type.
+     *
+     * Since different kinds of jobs get processed in a different way, we are able to generically process them based on
+     * their type after having registered a processor for them.
+     *
+     * @param jobClass class of the job that the GarbageCollector shall be able to handle
+     * @param queueProcessor function that takes care of processing the queue for this particular type
+     */
+    void registerQueueProcessor(Class<? extends GarbageCollectorJob> jobClass, QueueProcessor queueProcessor) {
+        this.queueProcessors.put(jobClass, queueProcessor);
+    }
+
+    /**
+     * This method allows us to register a {@link QueueConsolidator} for the given job type.
+     *
+     * Some jobs can be consolidated to consume less space in the queue by grouping them together or skipping them
+     * completely. While the consolidation of multiple jobs into fewer ones is optional and only required for certain
+     * types of jobs, this method allows us to generically handle this use case by registering a handler for the job
+     * class that supports this feature.
+     *
+     * @param jobClass class of the job that the GarbageCollector shall be able to handle
+     * @param queueConsolidator lambda that takes care of consolidating the entries in a queue
+     */
+    void registerQueueConsolidator(Class<? extends GarbageCollectorJob> jobClass, QueueConsolidator queueConsolidator) {
+        this.queueConsolidators.put(jobClass, queueConsolidator);
+    }
+
+    /**
+     * This method persists the changes of the garbage collector so IRI can continue cleaning up upon restarts.
+     *
+     * Since cleaning up the old database entries can take a long time, we need to make sure that it is possible to
+     * continue where we stopped. This method therefore writes the state of the queued jobs into a file that is read
+     * upon re-initialization of the GarbageCollector.
+     *
+     * @throws GarbageCollectorException if something goes wrong while writing the state file
+     */
+    void persistChanges() throws GarbageCollectorException {
+        try {
+            Files.write(
+                Paths.get(getStateFile().getAbsolutePath()),
+                () -> garbageCollectorJobs.entrySet().stream().flatMap(
+                    entry -> entry.getValue().stream()
+                ).<CharSequence>map(
+                    entry -> entry.getClass().getCanonicalName() + ";" + entry.serialize()
+                ).iterator()
+            );
```

Comment:
Just a question (since I am not a functional type of guy), do you find this readable?
Maybe it is just me not being down with functional programming?

Reply:
Yes, this is how it usually look :D